### PR TITLE
Drop support for EOL Python 2.6

### PR DESCRIPTION
**Pull request by hugovk**
For #883.

Requests and pip have already dropped Python 2.6.

This PR removes support for it, removes redundant code, and also does some modernising to use newer Python features.

### CODE REVIEW

File: src/urllib3/connection.py [hunk 1]
```diff
@@ -108,21 +104,10 @@ def __init__(self, *args, **kw):
         if six.PY3:  # Python 3
             kw.pop('strict', None)
 
-        # Pre-set source_address in case we have an older Python like 2.6.
-        self.source_address = kw.get('source_address')
```
Comment (SethMichaelLarson) [inline src/urllib3/connection.py]: I think we should keep this attribute on the `HTTPConnection` object, might be being used by others.

File: src/urllib3/util/ssl_.py [hunk 2]
```diff
@@ -115,10 +114,7 @@ def inet_pton(_, host):
 except ImportError:
     import sys
 
-    class SSLContext(object):  # Platform-specific: Python 2 & 3.1
-        supports_set_ciphers = ((2, 7) <= sys.version_info < (3,) or
-                                (3, 2) <= sys.version_info)
-
+    class SSLContext(object):  # Platform-specific: Python 2
```
[the remark this answers is not in the document]
Comment (hugovk) [reply src/urllib3/util/ssl_.py]: Reverted.

File: src/urllib3/connection.py (hunk 1 above)
Comment (hugovk) [reply src/urllib3/connection.py]: Kept.

**Comment by SethMichaelLarson** [comment]
Could you comment on how you found the occurrences for these optimizations? Was it a manual effort or is there a tool for the unittest improvements?

**Comment by hugovk** [comment]
Sure, for the set/dict/list literals, I used the code inspection (Code > Inspect Code...) in PyCharm (they have a lightweight [Community edition](https://www.jetbrains.com/pycharm/download/) and also free licences for the [full version for open source projects](https://www.jetbrains.com/opensource/)).

For the unit tests, I used the script at https://github.com/hugovk/upgrade-unittest which is just some sed commands, not a full AST parser, so you need to check the changes. There's also the https://github.com/jparise/flake8-assertive plugin, which I've just run and found some others which I've updated.

**Comment by SethMichaelLarson** [comment]
I'm approving these changes, just got to let CI pass now. :)

**Comment by SethMichaelLarson** [comment]
Thanks for doing this! :tada:

**Comment by nateprewitt** [comment]
As a quick note, we dropped support for 2.6 but haven’t merged the removal changes into master since that would be a breaking change for Requests 2.X.

We can probably go ahead and drop it, but we’ll need to make an announcement that it breaks now.

**Comment by hugovk** [comment]
You're welcome!

**Comment by pquentin** [comment]
Thank you again @hugovk, that was a quick and high-quality PR. :)

Would you be interested in working on [urllib3 async support](https://github.com/urllib3/urllib3/pull/1335)?

**Comment by hugovk** [comment]
Thanks, happy to help out, but I'm not yet really familiar with async.

**Comment by SethMichaelLarson** [comment]
Also @hugovk since you did the bulk of this work could you add a changelog entry for dropping Python 2.6 and add yourself to contributors list if you're not already there.

**Comment by hugovk** [comment]
@SethMichaelLarson Done! Please see https://github.com/urllib3/urllib3/pull/1431.

**Comment by pquentin** [comment]
@hugovk There are many ways to help even if you're not familiar with async yet! Removing Python 2.6 support was one of those things. :)

Another helpful thing to do is to improve the setup.py script: https://github.com/python-trio/trio/issues/567#issuecomment-408583539. There are many things to do but they are not documented yet.

Feel free to open an issue in https://github.com/python-trio/urllib3/ or send me an email at quentin.pradet@gmail.com if anything is unclear.

But no pressure!

**Pull request merged**